datafusion/core/src/physical_optimizer/replace_repartition_execs.rs

```diff
@@ -0,0 +1,802 @@
+///
+/// The algorithm flow is simply like this:
+/// 1. Visit nodes of the physical plan top-down and look for a SortExec node (if not found do nothing)
+/// 2. If `SortExec` is found, iterate over its children recursively until the leaf node
+///      or another `SortExec` and check if anything breaks the ordering between
+///      `RepartitionExec`s with multiple input partitions considered as they maintain input ordering
+///       because they're potentially changed with `SortPreservingRepartitionExec`s
+/// 2_1. If the ordering is broken somehow, do nothing
+/// 2_2. If the ordering is kept the same among all children continue
+/// 3. Start a transform loop bottom-up on `SortExec`s children
+/// 3_1. Check if the child is a `RepartitionExec` with multiple input partitions and if the ordering matches with the `SortExec`
+/// 3_1_1. If conditions match, replace the `RepartitionExec` with `SortPreservingRepartitionExec`
+/// 3_1_2. If conditions do not match, keep the plan as is.
+/// 4. End the bottom-up transformation loop
+/// 5. Check if the `SortExec` plan is changed somehow.
+/// 5_1. If changed remove the `SortExec` from the plan
+/// 5_2. If not changed then keep the plan as is.
+/// 6. Continue to the top-down transformation
+pub struct ReplaceRepartitionExecs {}
+
+impl ReplaceRepartitionExecs {
+    #[allow(missing_docs)]
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+/// Checks if repartition should be replaced.
+/// Takes parameters as one original `SortExec`, one `RepartitionExec` to compare,
+/// and a changed version of `SortExec`s input to be able to detect if the ordering will be preserved after the change
+fn is_repartition_matches_with_output_ordering(
+    sort_exec: &SortExec,
+    repartition: &RepartitionExec,
+) -> bool {
+    let orderings_matched = ordering_satisfy(
+        sort_exec.output_ordering(),
+        repartition.input().output_ordering(),
+        || repartition.input().equivalence_properties(),
+        || repartition.input().ordering_equivalence_properties(),
+    );
+    orderings_matched && !repartition.maintains_input_order()[0]
+}
+
+/// Creates a `SortPreservingRepartitionExec` from given `RepartitionExec`
+fn sort_preserving_repartition(
+    repartition: &RepartitionExec,
+) -> Arc<SortPreservingRepartitionExec> {
+    Arc::new(
+        SortPreservingRepartitionExec::try_new(
+            repartition.input().clone(),
+            repartition.partitioning().clone(),
+        )
+        .unwrap(),
+    )
+}
+
+fn does_plan_maintains_input_order(plan: &Arc<dyn ExecutionPlan>) -> bool {
+    return plan.maintains_input_order().iter().all(|flag| *flag);
+}
+
+/// Check the children nodes between two `SortExec`s if they all maintain the input ordering
+/// if all of the related children maintain, then it is a possibility to replace `RepartitionExec`s
+fn replace_sort_children_if_needed(
+    sort_exec: &SortExec,
+    plan: &Arc<dyn ExecutionPlan>,
+) -> Arc<dyn ExecutionPlan> {
```

Comment:
```suggestion
) -> Result<Arc<dyn ExecutionPlan>> {
```
It is better to return `Result` here, by this way we can remove `.unwrap`s in the function (Other than test and test util codes, there should be no `.unwrap`s (for most use cases) in the code).